src/properties/forest.c

```diff
@@ -0,0 +1,216 @@
+ *        out-tree, an in-tree or ignore edge directions. The respective
+ *        possible values are:
+ *        \c IGRAPH_OUT, \c IGRAPH_IN, \c IGRAPH_ALL. This argument is
+ *        ignored for undirected graphs.
+ * \return Error code:
+ *        \c IGRAPH_EINVMODE: invalid mode argument.
+ *
+ * Time complexity: At most O(|V|+|E|), the
+ * number of vertices plus the number of edges in the graph.
+ *
+ */
+igraph_error_t igraph_is_forest(const igraph_t *graph,igraph_bool_t *res, igraph_vector_t *roots , igraph_neimode_t mode) {
+    igraph_adjlist_t al;
+    igraph_vector_t visited;
+    igraph_integer_t visited_count=0;
+    igraph_integer_t vcount, ecount;
+
+    vcount = igraph_vcount(graph);
+    ecount = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_resize(roots,0));
+
+    /*By convention, a zero-vertex graph will be considered a forest.*/
+    if (vcount == 0) {
+        *res = 1;
+        return IGRAPH_SUCCESS;
+    }
+    // A forest can have maximum vcount-1 edges.
+    if (ecount > vcount - 1) {
+        *res = 0;
+        return IGRAPH_SUCCESS;
+    }
+    /*A single-vertex graph is a forest, provided it has no edges (checked in the previous if (..)) */
+    if (vcount == 1) {
+        *res = 1;
+        igraph_vector_push_back(roots,0);
+        return IGRAPH_SUCCESS;
+    }
+    /* Ignore mode for undirected graphs. */
+    if (! igraph_is_directed(graph)) {
+        mode = IGRAPH_ALL;
+    }
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &al, mode, IGRAPH_LOOPS, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+
+    *res = 1; /* assume success */
+    IGRAPH_CHECK(igraph_vector_init(&visited,vcount));
+    IGRAPH_FINALLY(igraph_vector_destroy, &visited);
+    igraph_integer_t i;
+
+    /* The main algorithm:
+     * Undirected Graph:- We add each unvisited vertex to the roots vector, and
+     * mark all other vertices that are reachable from it as visited.
+     *
+     * Directed Graph:- For each tree, the root is the node with no
+     * incoming/outgoing connections, depending on 'mode'. We add each vertex
+     * with zero degree to the roots vector and mark all other vertices that are
+     * reachable from it as visited.
+     *
+     * If all the vertices are visited exactly once, then the graph is a forest.
+     */
+
+    switch (mode) {
+    case IGRAPH_ALL:{
+
+        for(i =0; i<vcount;++i){
+            if (!VECTOR(visited)[i]) {
+                igraph_vector_push_back(roots,i);
+                IGRAPH_CHECK(igraph_i_is_forest_visitor(i,&visited, &al, &visited_count,res,mode));
+            }
+        }
+        break;
+    }
+    case IGRAPH_IN:
+    case IGRAPH_OUT: {
+        igraph_vector_int_t degree;
+
+        IGRAPH_CHECK(igraph_vector_int_init(&degree, 0));
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &degree);
+        IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), mode == IGRAPH_IN ? IGRAPH_OUT : IGRAPH_IN, /* loops = */ 1));
```

Comment:
@szhorvat has recently introduced a macro named `IGRAPH_REVERSE_MODE(...)`, which reverses the `mode` argument like you do it here. Let's start using it. This is not in the docs yet so that's why I'm telling now :)